charger/pulsatrix.go

```diff
@@ -0,0 +1,272 @@
+		ctx, cancel := context.WithTimeout(context.Background(), request.Timeout)
+		defer cancel()
+		messageType, message, err := c.conn.Read(ctx)
+		if err != nil {
+			c.log.ERROR.Println("error reading message:", err)
+			break
+		} else {
+			c.parseWsMessage(messageType, message)
+			c.updated = time.Now()
+		}
+	}
+	fmt.Println("wsReader stopped")
+	c.mutex.Lock()
+	if c.conn != nil {
+		c.conn.Close(websocket.StatusNormalClosure, "Reconnecting")
+		c.conn = nil
+	}
+	c.mutex.Unlock()
+	c.reconnectWs()
+
+}
+
+// wsWrite writes a message to the websocket
+func (c *PulsatrixCharger) wsWriter(message string) error {
+	if c.valid() && c.conn != nil {
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+		ctx, cancel := context.WithTimeout(context.Background(), request.Timeout)
+		defer cancel()
+		err := c.conn.Write(ctx, websocket.MessageText, []byte(message))
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("Waiting for reconnect")
+	}
+	return nil
+}
+
+// ParseWsMessage parses a message from the websocket
+func (c *PulsatrixCharger) parseWsMessage(messageType websocket.MessageType, message []byte) {
+	if messageType == websocket.MessageText {
+		b := bytes.ReplaceAll(message, []byte(":NaN"), []byte(":null"))
+		idx := bytes.IndexByte(b, '{')
+		c.handleError(json.Unmarshal(b[idx:], c))
+
+	}
+}
+
+// Heartbeat sends a heartbeat to the pulsatrix SECC
+func (c *PulsatrixCharger) heartbeat() {
+	for range time.Tick(3 * time.Minute) {
+		c.handleError(c.Enable(c.enState))
+
+	}
+}
+
+func (c *PulsatrixCharger) valid() bool {
+	return time.Since(c.updated) < 30 * time.Second
+}
+
+func (c *PulsatrixCharger) handleError(err error) {
+	if err != nil {
+		c.log.ERROR.Println(err)
+	}
+}
+
+// Status implements the api.Charger interface
+func (c *PulsatrixCharger) Status() (api.ChargeStatus, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return api.ChargeStatusString(c.VehicleStatus)
+}
+
+// Enabled implements the api.Charger interface
+func (c *PulsatrixCharger) Enabled() (bool, error) {
+	return verifyEnabled(c, c.enState)
+}
+
+// Enable implements the api.Charger interface
+func (c *PulsatrixCharger) Enable(enable bool) error {
+	c.enState = enable
+	return c.wsWriter("setEnabled\n" + strconv.FormatBool(c.enState))
+}
+
+// MaxCurrent implements the api.CurrentLimiter interface
+func (c *PulsatrixCharger) MaxCurrent(current int64) error {
+	return c.MaxCurrentMillis(float64(current))
+}
+
+// MaxCurrentMillis implements the api.ChargerEx interface
+func (c *PulsatrixCharger) MaxCurrentMillis(current float64) error {
+	c.signaledAmp = current
+	if !c.enState && current > 0 {
+		c.handleError(c.Enable(true))
+	}
+	return c.wsWriter("setCurrentLimit\n" + strconv.FormatFloat(current, 'f', 10, 64))
+}
+
+// GetMaxCurrent implements the api.CurrentGetter interface
+func (c *PulsatrixCharger) GetMaxCurrent() (float64, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if !c.valid() {
+		c.handleError(c.Enable(false))
```

Comment:
dito

Reply:
And again: either log or return the error. not both.

Reply:
as I said above, those are 2 different errors:

```suggestion
		c.handleError(c.Enable(false))
		return 0, fmt.Errorf(api.ErrOutdated.Error())
```

the fist line (the log) handles the error if the c.Enable function call fails
the second one (the return) is for the outdated data and a different error